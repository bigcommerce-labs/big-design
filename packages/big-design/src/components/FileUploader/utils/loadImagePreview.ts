export const loadImagePreview = (file: File): Promise<string | null> => {
  if (!file.type.startsWith('image/')) {
    return Promise.resolve(null);
  }

  return new Promise((resolve, reject) => {
    const reader = new FileReader();

    reader.onload = (event) => {
      // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
      const imageUrl = event.target?.result as string;

      resolve(imageUrl);
    };

    reader.onerror = (error) => {
      // eslint-disable-next-line @typescript-eslint/prefer-promise-reject-errors
      reject(error);
    };

    reader.readAsDataURL(file);
  });
};

export const getImagesPreview = async (files: File[]) =>
  files.reduce(async (prevPromise, file, idx) => {
    const prevResult = await prevPromise;
    const imageUrl = await loadImagePreview(file);

    return { ...prevResult, [idx]: imageUrl };
  }, Promise.resolve<Record<number, string | null>>({}));
