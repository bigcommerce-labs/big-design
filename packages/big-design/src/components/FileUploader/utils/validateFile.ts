import { ValidatorConfig } from '../types';

export const validateFileFormat = (file: File | DataTransferItem, fileFormats = '') => {
  const allowedFormats = fileFormats.split(',').map((format) => format.trim());
  const fileExtension = file.type.split('/').pop()?.toLowerCase() ?? '';

  return allowedFormats.some((format) => {
    if (format === '*' || format === '.*') {
      return true;
    } else if (format.startsWith('.')) {
      return format === `.${fileExtension}`;
    } else if (
      format.startsWith('audio/') ||
      format.startsWith('video/') ||
      format.startsWith('image/')
    ) {
      const mediaTypePrefix = format.split('/')[0];

      return file.type.startsWith(`${mediaTypePrefix}/`);
    }

    return true;
  });
};

export const validateFiles = (files: File[], validators: ValidatorConfig[]) =>
  files.flatMap((file, idx) =>
    validators
      .filter(({ validator }) => !validator(file))
      .map(({ message, type }) => ({ file, fileIdx: idx, message, type })),
  );
