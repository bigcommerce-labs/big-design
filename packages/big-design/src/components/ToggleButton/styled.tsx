import { theme as defaultTheme } from '@bigcommerce/big-design-theme';
import styled from 'styled-components';

interface StyledToggleButtonProps {
  selected: boolean;
}

export const StyledToggleButton = styled('button')<StyledToggleButtonProps>`
  background: ${({ selected, theme }) => (selected ? theme.colors.primary20 : theme.colors.white)};
  border: 1px solid ${({ selected, theme }) => (selected ? theme.colors.primary30 : theme.colors.secondary30)};
  border-radius: ${({ theme }) => theme.borderRadius.normal};
  color: ${({ selected, theme }) => (selected ? theme.colors.primary60 : theme.colors.secondary60)};
  cursor: pointer;
  font-size: ${({ theme }) => theme.typography.fontSize.medium};
  line-height: ${({ theme }) => theme.lineHeight.xLarge};
  margin: 0;
  outline: none;
  padding: 0 ${({ theme }) => theme.spacing.medium};
  position: relative;

  &[disabled] {
    background: ${({ selected, theme }) => (selected ? theme.colors.secondary10 : theme.colors.white)};
    border: 1px solid ${({ theme }) => theme.colors.secondary30};
    color: ${({ theme }) => theme.colors.secondary40};
    cursor: not-allowed;
  }

  &:not([disabled]):hover,
  &:not([disabled]):active {
    background: ${({ selected, theme }) => (selected ? theme.colors.primary20 : theme.colors.primary10)};
    color: ${({ selected, theme }) => (selected ? theme.colors.primary60 : theme.colors.primary50)};
  }

  &:not([disabled]):focus {
    box-shadow: 0 0 0 4px ${({ theme }) => theme.colors.primary20};
  }
`;

StyledToggleButton.defaultProps = { theme: defaultTheme };
