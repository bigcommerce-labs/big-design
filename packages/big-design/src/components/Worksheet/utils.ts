import { Cell, WorksheetError, WorksheetItem } from './types';

export const mergeCells = <T extends WorksheetItem>(oldCells: Cell<T>[], newCells: Cell<T>[]) =>
  newCells.reduce(
    (accum, newCell) => {
      const index = oldCells.findIndex(
        (oldCell) => oldCell.rowIndex === newCell.rowIndex && oldCell.columnIndex === newCell.columnIndex,
      );

      if (index > -1) {
        accum[index] = newCell;

        return accum;
      }

      return accum.concat(newCell);
    },
    [...oldCells], // Note: returns a new array every time
  );

export const deleteCells = <T extends WorksheetItem>(oldCells: Cell<T>[], newCells: Cell<T>[]) =>
  oldCells.filter(
    (oldCell) =>
      !newCells.find((newCell) => newCell.columnIndex === oldCell.columnIndex && newCell.rowIndex === oldCell.rowIndex),
  );

export const editedRows = <T extends WorksheetItem>(editedCells: Cell<T>[], rows: T[]) =>
  editedCells.reduce<T[]>((accum, { rowIndex }) => {
    const row = rows[rowIndex];

    // Check to see if the row already exists in accum
    if (accum.find((editedRow) => editedRow === row)) {
      return accum;
    } else {
      // Only append new rows
      return [...accum, row];
    }
  }, []);

export const invalidRows = <T extends WorksheetItem>(invalidCells: Cell<T>[], rows: T[]) => {
  const mapObj = new Map();

  // Create Map with each row and append errors per row
  invalidCells.forEach(({ rowIndex, hash }) => {
    const row = rows[rowIndex];
    if (mapObj.has(row)) {
      const errors = mapObj.get(row);
      mapObj.set(row, new Set([...errors, hash]));
    } else {
      mapObj.set(row, new Set([hash]));
    }
  });

  return Array.from(mapObj).map<WorksheetError<T>>(([item, errors]) => ({
    item,
    errors: Array.from(errors),
  }));
};
