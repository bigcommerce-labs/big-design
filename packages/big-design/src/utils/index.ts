export * from './typedMemo';
export * from './uniqueId';
export * from './warning';
