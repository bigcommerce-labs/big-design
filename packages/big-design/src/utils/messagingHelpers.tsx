import { CheckCircleIcon, ErrorIcon, InfoIcon, WarningIcon } from '@bigcommerce/big-design-icons';
import { ThemeInterface } from '@bigcommerce/big-design-theme';
import React, { ComponentPropsWithoutRef } from 'react';
import { css } from 'styled-components';

import { ButtonProps } from '../components/Button';
import { LinkProps } from '../components/Link';
import { MarginProps } from '../mixins/margins';

export interface SharedMessagingProps extends ComponentPropsWithoutRef<'div'> {
  actions?: MessageAction[];
  header?: string;
  messages: MessageItem[];
  type?: MessagingType;
  onClose?(): void;
}

export type MessagingType = 'success' | 'error' | 'warning' | 'info';

export interface MessageItem {
  text: string;
  link?: MessageLinkItem;
}

export type MessageLinkItem = Pick<LinkProps, 'external' | 'href' | 'target'> & {
  text: string;
};

export interface MessageAction
  extends Omit<ButtonProps, 'children' | 'mobileWidth' | 'variant' | keyof MarginProps> {
  text: string;
  variant?: Exclude<ButtonProps['variant'], 'primary'>;
}

export const getMessagingIcon = (type: MessagingType, condensed?: boolean) => {
  const size = condensed ? 'large' : 'xLarge';

  switch (type) {
    case 'success':
      return <CheckCircleIcon color="success" size={size} />;

    case 'error':
      return <ErrorIcon color="danger" size={size} />;

    case 'warning':
      return <WarningIcon color="warning50" size={size} />;

    case 'info':
      return <InfoIcon color="primary60" size={size} />;
  }
};

export const getBorderStyle = (type: MessagingType, theme: ThemeInterface) => css`
  ${type === 'success' &&
  css`
    border-left: ${theme.spacing.xxSmall} solid ${theme.colors.success};
  `};

  ${type === 'error' &&
  css`
    border-left: ${theme.spacing.xxSmall} solid ${theme.colors.danger};
  `};

  ${type === 'warning' &&
  css`
    border-left: ${theme.spacing.xxSmall} solid ${theme.colors.warning50};
  `};

  ${type === 'info' &&
  css`
    border-left: ${theme.spacing.xxSmall} solid ${theme.colors.primary60};
  `};
`;

export const getActionVariant = (variant: ButtonProps['variant']) =>
  variant === 'primary' ? 'secondary' : variant;
