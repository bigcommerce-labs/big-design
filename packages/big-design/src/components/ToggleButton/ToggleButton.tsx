import React from 'react';

import { StyledToggleButton } from './styled';

export interface ToggleButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  selected: boolean;
}

export const ToggleButton: React.FC<ToggleButtonProps> = React.memo(({ className, style, ...props }) => {
  const { selected } = props;

  return (
    <StyledToggleButton aria-pressed={selected} selected={selected} {...props}>
      {props.children}
    </StyledToggleButton>
  );
});
