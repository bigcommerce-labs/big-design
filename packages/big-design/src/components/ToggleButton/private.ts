import { ToggleButtonProps as _ToggleButtonProps } from './ToggleButton';
export { ToggleButton } from './ToggleButton';
export type ToggleButtonProps = _ToggleButtonProps;
