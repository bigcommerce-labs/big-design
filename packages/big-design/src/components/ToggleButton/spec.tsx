import { fireEvent, render } from '@test/utils';
import 'jest-styled-components';
import React from 'react';

import { ToggleButton } from './ToggleButton';

test('render selected and enabled', () => {
  const { container } = render(<ToggleButton selected={true}></ToggleButton>);

  expect(container.firstChild).toMatchSnapshot();
});

test('render unselected and enabled', () => {
  const { container } = render(<ToggleButton selected={false}></ToggleButton>);

  expect(container.firstChild).toMatchSnapshot();
});

test('render selected and disabled', () => {
  const { container } = render(<ToggleButton selected={true} disabled></ToggleButton>);

  expect(container.firstChild).toMatchSnapshot();
});

test('render unselected and disabled', () => {
  const { container } = render(<ToggleButton selected={false} disabled></ToggleButton>);

  expect(container.firstChild).toMatchSnapshot();
});

test('does not forward styles', () => {
  const { container } = render(<ToggleButton className="test" selected={true} style={{ backgroundColor: 'orange' }} />);

  expect(container.getElementsByClassName('test').length).toBe(0);
  expect(container.style.backgroundColor).toBe('');
});

test('triggers onClick', () => {
  const onClick = jest.fn();

  const { container } = render(<ToggleButton selected={true} onClick={onClick}></ToggleButton>);
  const button = container.firstChild as HTMLElement;

  fireEvent.click(button);

  expect(onClick).toHaveBeenCalled();
});
