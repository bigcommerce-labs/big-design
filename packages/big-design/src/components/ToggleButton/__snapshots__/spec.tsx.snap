// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render selected and disabled 1`] = `
.c0 {
  background: #DBE3FE;
  border: 1px solid #9EB3FC;
  border-radius: 0.25rem;
  color: #0B38D9;
  cursor: pointer;
  font-size: 1rem;
  line-height: 2rem;
  margin: 0;
  outline: none;
  padding: 0 1rem;
  position: relative;
}

.c0[disabled] {
  background: #F6F7FC;
  border: 1px solid #D9DCE9;
  color: #B4BAD1;
  cursor: not-allowed;
}

.c0:not([disabled]):hover,
.c0:not([disabled]):active {
  background: #DBE3FE;
  color: #0B38D9;
}

.c0:not([disabled]):focus {
  box-shadow: 0 0 0 4px #DBE3FE;
}

<button
  aria-pressed="true"
  class="c0"
  disabled=""
/>
`;

exports[`render selected and enabled 1`] = `
.c0 {
  background: #DBE3FE;
  border: 1px solid #9EB3FC;
  border-radius: 0.25rem;
  color: #0B38D9;
  cursor: pointer;
  font-size: 1rem;
  line-height: 2rem;
  margin: 0;
  outline: none;
  padding: 0 1rem;
  position: relative;
}

.c0[disabled] {
  background: #F6F7FC;
  border: 1px solid #D9DCE9;
  color: #B4BAD1;
  cursor: not-allowed;
}

.c0:not([disabled]):hover,
.c0:not([disabled]):active {
  background: #DBE3FE;
  color: #0B38D9;
}

.c0:not([disabled]):focus {
  box-shadow: 0 0 0 4px #DBE3FE;
}

<button
  aria-pressed="true"
  class="c0"
/>
`;

exports[`render unselected and disabled 1`] = `
.c0 {
  background: #FFFFFF;
  border: 1px solid #D9DCE9;
  border-radius: 0.25rem;
  color: #5E637A;
  cursor: pointer;
  font-size: 1rem;
  line-height: 2rem;
  margin: 0;
  outline: none;
  padding: 0 1rem;
  position: relative;
}

.c0[disabled] {
  background: #FFFFFF;
  border: 1px solid #D9DCE9;
  color: #B4BAD1;
  cursor: not-allowed;
}

.c0:not([disabled]):hover,
.c0:not([disabled]):active {
  background: #F0F3FF;
  color: #2852EB;
}

.c0:not([disabled]):focus {
  box-shadow: 0 0 0 4px #DBE3FE;
}

<button
  aria-pressed="false"
  class="c0"
  disabled=""
/>
`;

exports[`render unselected and enabled 1`] = `
.c0 {
  background: #FFFFFF;
  border: 1px solid #D9DCE9;
  border-radius: 0.25rem;
  color: #5E637A;
  cursor: pointer;
  font-size: 1rem;
  line-height: 2rem;
  margin: 0;
  outline: none;
  padding: 0 1rem;
  position: relative;
}

.c0[disabled] {
  background: #FFFFFF;
  border: 1px solid #D9DCE9;
  color: #B4BAD1;
  cursor: not-allowed;
}

.c0:not([disabled]):hover,
.c0:not([disabled]):active {
  background: #F0F3FF;
  color: #2852EB;
}

.c0:not([disabled]):focus {
  box-shadow: 0 0 0 4px #DBE3FE;
}

<button
  aria-pressed="false"
  class="c0"
/>
`;
